[darHH] iP

darHH opened the pull request:
# Dar Chatbot

> "Dar Chatbot has changed my life." - God ([source](http://trust-me-bro.com/))

Dar Chatbot gives you the company and help you never knew you needed. Dar can be your:
- Personal Assistant
- Best Friend 
- Advisor 
- Therapist

All you need to do is:
1. Download it from [here](url)
2. Start talking to it

Nothing in this world comes free, **except** for Dar Chatbot!

## Features
- [x] Managing tasks
- [ ] Cure Depression (coming soon) 
- [ ] Earn Money (coming soon)

---

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

darHH committed:
Added greeting and exit message

darHH committed:
Echo function implemented

darHH committed:
List input and output added

darHH committed:
Created Task class

darHH committed:
Change location of Task file

darHH committed:
Added instruction set and integrated Task class to Dar

darHH committed:
Implemented list function

darHH committed:
Used Consumer interface

darHH committed:
Implemented mark and unmark functions

darHH committed:
ToDo class done, added unknowninputmessage, created empty deadline and event classes

darHH committed:
Implemented Deadline class

darHH committed:
Implemented Event and fixed all bugs

darHH committed:
Created automated tests

darHH committed:
Error handling for invalid todo, deadline, event, and unknown instruction

darHH committed:
Implemented delete function with updated test cases

darHH committed:
Created Storage class and implemented reading and writing from/to Storage to/from Task class and subclasses

darHH committed:
Implemented read on startup and rewrite on close (bye)

darHH committed:
Every action that affects task has save to data.

darHH committed:
Rewrote trivial testing (yet to implement for save function)

darHH committed:
Merge branch 'branch-Level-7'

darHH committed:
Implemented Dates and Times for deadline task

darHH committed:
Fixed a printing bug

darHH committed:
Formatted tostring for deadline and added erorr messages for invalid date and time format

darHH committed:
Merge branch 'branch-Level-8'

darHH committed:
Implemented "Parser" Class

darHH committed:
Add UI class

darHH committed:
Add CommandManager class

darHH committed:
Update syntax

darHH committed:
Merge branch 'A-MoreOOP'

darHH committed:
divide classes into 3 packages; dar, task, command

darHH committed:
Merge remote-tracking branch 'origin/add-gradle-support'

darHH committed:
Add and test gradle

darHH committed:
Add StorageTest

darHH committed:
Add DeadlineTest

DeadlineTest contains 2 tests that checks if the date and time of a deadline task is correctly saved

darHH committed:
Merge branch 'A-JUnit'

darHH committed:
Change jar file name from duke to dar

darHH committed:
Add Javadoc comments for all classes (most methods)

darHH committed:
Remove presence of Javadoc comment check in Checkstyle.xml

darHH committed:
Fix all Checkstyle main errors

darHH committed:
Add find tasks command

darHH committed:
Merge branch 'branch-A-JavaDoc'

darHH committed:
Merge branch 'branch-A-CodingStandard'

darHH committed:
Merge branch 'branch-Level-9'

darHH committed:
Fix checkstyleMain errors

philbertshea commented on src/main/java/dar/Dar.java:
```diff
@@ -0,0 +1,67 @@
+
+import command.CommandManager;
+
+public class Dar {
```
Maybe you could add a Javadoc comment for this class?

philbertshea commented on src/main/java/dar/Storage.java:
```diff
@@ -0,0 +1,80 @@
+     * If the parent directory or data file does not exist, it is created.
+     * If an error occurs during file creation, an error message is displayed.
+     */
+    private void doesFileExist() {
```
Maybe you could rename this to something like createFileAndDirectoryIfNotExists?

philbertshea commented on src/main/java/dar/Storage.java:
```diff
@@ -0,0 +1,80 @@
+     *
+     * @param tasks The list of tasks to save.
+     */
+    public void saveTasks(ArrayList<Task> tasks) {
```
Great naming for the saveTasks and loadTasks methods! Right to the point.

philbertshea commented on src/main/java/task/Deadline.java:
```diff
@@ -0,0 +1,110 @@
+        if (currentNumber.length() > 0) {
+            numberSequence.add(currentNumber.toString());
+        }
+        boolean validDateOrTime = false;
```
Maybe rename it to isValidDateOrTime?

philbertshea commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,84 @@
+        this.taskNumber = newTaskNumber;
+    }
+
+    public void totalTasksMinusOne() {
```
Maybe rename this to decrementTotalTasksCount?

philbertshea reviewed (commented):
Great work in general! 💯 Just some nitpicking here and there to be honest. Keep up the great work.

darHH committed:
Improvement to Java coding standard (feedback from TA)

darHH committed:
Fic checkstyle errors

darHH committed:
Merge branch 'A-CheckStyle'

darHH committed:
Add Javadoc comments for class declartaion and constructor

darHH committed:
Add JavaFX GUI, dar echoes only

darHH committed:
Remove print messages and return Strings for getResponse() in Dar class for GUI to work with chatbot

darHH committed:
Modify deadline to handle invalid date/time format better

darHH committed:
Remove text-based UI test

darHH committed:
Fix checkstyleMain errors

darHH committed:
Merge branch 'branch-Level-10'

sunpterodactyl commented on src/main/java/command/CommandManager.java:
```diff
@@ -0,0 +1,222 @@
+        response.append("(Numbers represent that task's number, for deleting and marking etc.)\n\n");
+
+        matchWord = matchWord.toLowerCase();
+        boolean found = false;
```
Maybe add a prefix to the boolean e.g. isFound?

sunpterodactyl commented on src/main/java/gui/DialogBox.java:
```diff
@@ -0,0 +1,61 @@
+import javafx.scene.layout.HBox;
+
+/**
+ * Represents a dialog box consisting of an ImageView to represent the speaker's face
```
Great work on the javadoc descriptions overall!

sunpterodactyl commented on src/main/java/command/CommandManager.java:
```diff
@@ -0,0 +1,222 @@
+     * <p>
+     * If the task list is empty, return a String message indicating no tasks.
+     * <p>
+     * This method does not return anything.
```
The Javadoc mentions listTasks does not return anything. Maybe you forgot to update this?

sunpterodactyl commented on src/main/java/dar/Ui.java:
```diff
@@ -0,0 +1,52 @@
+    /**
+     * Displays greeting message
+     * <p>
+     * This method does not return anything.
```
I think you forgot to update the Javadocs to reflect the return values in this class?

sunpterodactyl reviewed (commented):
Well done :raised_hands:  - just a few minor comments from me. Keep up the amazing work!

darHH committed:
The saveDateTime method in Deadline class handles both extracting numeric sequences and parsing them into date/time values, making it less readable and harder to maintain.

Mixing extraction and parsing logic in one method violates Single Responsibility Principle (SRP) and reduces code clarity.

Let’s extract extractNumericSequences to handle sequence extraction and parseDateTime to validate and parse dates/times.

Separating these concerns improves modularity, testability, and maintainability, making future modifications easier.

darHH committed:
The Storage and Parser classes assume valid input but do not enforce constraints, leading to potential issues such as null inputs.

Without assertions, invalid states may go undetected, making debugging difficult and potentially causing undefined behavior.

Let’s add assertions to:
Storage: Verify that the file path is not null and that serialized/deserialized tasks are valid.
Parser: Ensure that input text is not null

These assertions improve debugging, code safety, and data integrity by detecting invalid states early.

darHH committed:
The Task class does not enforce constraints on taskNumber and totalTasks, allowing negative or invalid values, which can lead to inconsistencies and bugs.

Let’s add assertions to:
Constructor: Ensure taskNumber is always positive after initialization.
Task count management: Prevent totalTasks from being negative after incrementing or decrementing.
Task number updates: Allow only positive values.

These assertions improve data integrity and debugging by detecting invalid task states early while remaining performance-neutral in production.

darHH committed:
The Deadline and Event constructors handled both task initialization and description parsing, making them less readable and harder to maintain.

Mixing object construction with parsing logic violates the Single Responsibility Principle (SRP) and makes future modifications more complex.

Let’s extract parseDescription in both classes to handle parsing separately.
Deadline: Extracts the description and deadline details from the input.
Event: Extracts the event description and time range from “from” and “to”.

This refactoring improves code readability, maintainability, and separation of concerns, making the constructors more focused and easier to extend.

darHH committed:
Command manager: replace loops with stream operations in task listing and searching

The CommandManager class uses traditional for-loops and StringBuilder concatenation to handle task listing and searching operations.

Let's replace the loops with stream operations in listTasks() and findTasks() methods. Using streams here improves code readability.

The mark/unmark and add operations remain unchanged as their current  implementation is more straightforward and easier to maintain.

darHH committed:
Merge pull request #1 from darHH/branch-A-Assertions

merge Branch-A-Assertions

darHH committed:
Merge pull request #2 from darHH/branch-A-CodeQuality

merge branch-A-CodeQuality

darHH committed:
Merge branch 'master' into branch-A-Streams

darHH committed:
Merge pull request #3 from darHH/branch-A-Streams

merge branch-A-Streams

darHH committed:
Merge branch 'master' of https://github.com/darHH/ip

darHH committed:
Merge branch 'branch-A-Streams' of https://github.com/darHH/ip into branch-A-Streams

darHH committed:
Merge branch 'branch-A-Streams'

darHH committed:
Create main.yml

darHH committed:
Create gradle.yml

darHH committed:
Merge branch 'master' of https://github.com/darHH/ip

darHH committed:
Fix checkstyle errors

darHH committed:
Implement detect duplicates

darHH committed:
Merge branch 'branch-C-DetectDuplicates'

darHH committed:
Implement a method to sort all deadline tasks in chronological order.

Tasks are filtered from the task list, sorted by date and time, and formatted for display.
Returns a message if no deadlines are found.

darHH committed:
update GUI: Improve dialog box and update chat images

darHH committed:
Change chat background image to farm theme

darHH committed:
Deadline.java: Handle invalid deadline command correctly

darHH committed:
Add JUnit tests for addTodo and markTask methods

darHH committed:
Deadline.java: Use constants for date and time patters to avoid magic strings

darHH committed:
Fix checkstyle errors

darHH committed:
Add user guide and screenshot

darHH committed:
Update user guide

darHH committed:
Update user guide

darHH committed:
Improve user guide

darHH committed:
Improve user guide

darHH committed:
Fix JavaFX app for windows and linux

darHH committed:
Fix JavaFX

darHH committed:
Fix test case

darHH committed:
Fix JavaFX
